src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Utilities/NullableHelpers/NullableHelpers.cs

```diff
@@ -0,0 +1,109 @@
+                hasReference = true;
+
+                // foreach statements are handled special because the iterator is not assignable, so the elementtype 
+                // annotation is accurate for determining if the loop declaration has a reference that allows the symbol
+                // to be null
+                if (reference is IForEachLoopOperation forEachLoop)
+                {
+                    var foreachInfo = semanticModel.GetForEachStatementInfo((CommonForEachStatementSyntax)forEachLoop.Syntax);
+
+                    if (foreachInfo.ElementType is null)
+                    {
+                        continue;
+                    }
+
+                    // Use NotAnnotated here to keep both Annotated and None (oblivious) treated the same, since
+                    // this is directly looking at the annotation and not the flow state
+                    if (foreachInfo.ElementType.NullableAnnotation != NullableAnnotation.NotAnnotated)
+                    {
+                        return true;
+                    }
+
+                    continue;
+                }
+
+                var syntax = reference is IVariableDeclaratorOperation variableDeclarator
+                    ? variableDeclarator.GetVariableInitializer().Value.Syntax
+                    : reference.Syntax;
+
+                var typeInfo = semanticModel.GetTypeInfo(syntax);
+
+                if (typeInfo.Nullability.FlowState == NullableFlowState.MaybeNull)
+                {
+                    return true;
+                }
+            }
+
+            return hasReference ? (bool?)false : null;
+        }
+
+        private static bool IsSymbolReferencedByOperation(IOperation operation, ISymbol symbol, bool allowNullInitializer)
+            => operation switch
+            {
+                ILocalReferenceOperation localReference => localReference.Local.Equals(symbol),
+                IParameterReferenceOperation parameterReference => parameterReference.Parameter.Equals(symbol),
+                IAssignmentOperation assignment => assignment is not IDeconstructionAssignmentOperation && IsSymbolReferencedByOperation(assignment.Target, symbol, allowNullInitializer: false),
+                IForEachLoopOperation loopOperation => IsSymbolReferencedByOperation(loopOperation.LoopControlVariable, symbol, allowNullInitializer: true),
+                ITupleOperation tupleOperation => tupleOperation.Elements.Any(element => IsSymbolReferencedByOperation(element, symbol, allowNullInitializer: false)),
+
+                // We want to be explicit about if we allow the variable to have a null variable initializer. The use case for now is around foreach loops 
+                // where the LoopControlVariable of the operation will have a null initializer.
+                IVariableDeclaratorOperation variableDeclarator => (allowNullInitializer || variableDeclarator.GetVariableInitializer() != null) && variableDeclarator.Symbol.Equals(symbol),
```

Comment:
fwiw, i find this condition super hard to understand :)

Reply:
If an `IVariableDeclaratorOperation` doesn't have a variable initializer, it is an unassigned variable right? In which case the operation shouldn't contribute to determining if a null value could be assigned. It's uninitialized. 

However, `IVariableDeclaratorOperation` always have a null variable initializer. I'm sure there's a very reasonable explanation for this, but in practice it means we still want to consider those values as assigned. The actual assignments would be generated code in the enumeration (if my understanding is correct).

Reply:
I'm still confused. :smile: Did you mean to write IVariableDeclaratorOperation in both those paragraphs?

Reply:
Oops, "However, IVariableDeclaratorOperation *in a foreach* always have a null variable initializer." is the correct thing. The range variable initializer is null

Reply:
This looks a lot simpler to me!